lib/FrameManager.js

```diff
@@ -112,16 +115,36 @@ class FrameManager extends EventEmitter {
       this._removeFramesRecursively(frame);
   }
 
-  _onExecutionContextCreated(context) {
-    const frameId = context.auxData && context.auxData.isDefault ? context.auxData.frameId : null;
+  _onExecutionContextCreated(contextPayload) {
+    const context = new ExecutionContext(this._client, contextPayload.id, this.createJSHandle.bind(this, contextPayload.id));
+    this._contextIdToContext.set(contextPayload.id, context);
+
+    const frameId = contextPayload.auxData && contextPayload.auxData.isDefault ? contextPayload.auxData.frameId : null;
     const frame = this._frames.get(frameId);
     if (!frame)
       return;
-    frame._defaultContextId = context.id;
+    frame._context = context;
     for (const waitTask of frame._waitTasks)
       waitTask.rerun();
   }
 
+  _onExecutionContextDestroyed(contextPayload) {
+    this._contextIdToContext.delete(contextPayload.id);
+  }
+
+  /**
+   * @param {string} contextId
+   * @param {*} remoteObject
+   * @return {!JSHandle}
+   */
+  createJSHandle(contextId, remoteObject) {
+    const context = this._contextIdToContext.get(contextId);
+    console.assert(context, 'INTERNAL ERROR: missing context with id = ' + contextId);
+    if (remoteObject.subtype === 'node')
+      return new ElementHandle(context, this._client, remoteObject, this._mouse, this._touchscreen);
```

Comment:
Can we check if it is an element here, instead of just a node?

Reply:
We can't: it requires either an async hop to the page, or breaking change to the protocol. We'd be very unhappy with both.